python/hsml/engine/model_engine.py

```diff
@@ -99,42 +99,27 @@ def _copy_hopsfs_model(self, model_path, dataset_model_version_path):
             _, file_name = os.path.split(path)
             self._dataset_api.copy(path, dataset_model_version_path + "/" + file_name)
 
-    def _upload_local_model_folder(self, model_path, dataset_model_version_path):
+    def _upload_local_model_folder(
+        self, local_model_path, model_version, dataset_model_name_path
+    ):
         archive_out_dir = None
+        uploaded_archive_path = None
         try:
             archive_out_dir = tempfile.TemporaryDirectory(dir=os.getcwd())
-            archive_path = util.compress(archive_out_dir.name, model_path)
-            self._dataset_api.upload(archive_path, dataset_model_version_path)
+            archive_path = util.compress(
+                archive_out_dir.name, str(model_version), local_model_path
+            )
+            uploaded_archive_path = (
+                dataset_model_name_path + "/" + os.path.basename(archive_path)
+            )
+            self._dataset_api.upload(archive_path, dataset_model_name_path)
+            self._dataset_api.unzip(uploaded_archive_path, block=True, timeout=600)
         except RestAPIError:
             raise
         finally:
             if archive_out_dir is not None:
                 archive_out_dir.cleanup()
-
-        extracted_archive_path = (
-            dataset_model_version_path + "/" + os.path.basename(archive_path)
-        )
-
-        self._dataset_api.unzip(extracted_archive_path, block=True, timeout=480)
-
-        self._dataset_api.rm(extracted_archive_path)
-
-        extracted_model_dir = (
-            dataset_model_version_path
-            + "/"
-            + os.path.basename(archive_path[: archive_path.index(".")])
-        )
-
-        # Observed that when decompressing a large folder and directly moving the files sometimes caused filesystem exceptions
-        time.sleep(5)
-
-        for artifact in os.listdir(model_path):
-            _, file_name = os.path.split(artifact)
-            self._dataset_api.move(
-                extracted_model_dir + "/" + file_name,
-                dataset_model_version_path + "/" + file_name,
-            )
-        self._dataset_api.rm(extracted_model_dir)
+            self._dataset_api.rm(uploaded_archive_path)
 
     def _set_model_version(
```

Comment:
I think we should add an endpoint in the REST API to get the last version available, in another PR. This endpoint could be use to decide the next version (last_version + 1) and when printing the available versions in the UI (range(0, last_version)).